config/emacs/config.org
#+TITLE: My Doom Emacs config
#+EMAIL: shanyouli6@gmail.com
#+DATA: 20 Nov 2020
#+AUTHOR: Syl
#+STARTUP: overview
+ [[https://github.com/hlissner/doom-emacs/blob/develop/docs/getting_started.org][doom-emacs/getting_started.org at develop · hlissner/doom-emacs]]
* Emacs
Emacs or EMACS(Editor MACroS) is a family of text editors that are characterized by
their extensibility. The manual for the most widely used variant, GNU Emacs, describes
it as "the extensible, customizable, self-documenting, real-time display editor".
Develoment of the first Emacs began in the mid-1970s, and work on its direct
descendant, GNU Emacs, continues actively as of 2020.

* Rudimentary configuration
Make this file run (slightly) faster with lexical binding ( see [[https://nullprogram.com/blog/2016/12/22/][this blog post]] for more info).
** config.el
#+begin_src emacs-lisp :tangle "config.el"
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;;; If you use NIXOS yourself to generate the configuration, you may need to load it.
(load! "config.init.el")
#+end_src
** packages.el
#+begin_src emacs-lisp :tangle "packages.el"
;;; $DOOMDIR/package.el -*- lexical-binding: t; -*-

;;; Load NIXOS automatically generated configuration.
(load! "packages.init.el")

(unpin! doom-modeline)
#+end_src

* set my personal Information
It is useful to have some basic personal information.
#+begin_src emacs-lisp :tangle "config.el"
(setq user-full-name "Shanyou Li"
      user-mail-address "shanyouli6@gmail.com")

(defun init-default-frame (&optional frame wratio hratio)
  (interactive)
  (let ((x-width (or (alist-get 'width default-frame-alist)
                     (truncate (- (* (x-display-pixel-width)
                                     (or wratio 0.5)) 20))))
        (y-height (or (alist-get 'height default-frame-alist)
                      (truncate (* (x-display-pixel-height)
                                   (or hratio 0.5))))))
    (set-frame-size (or frame (selected-frame)) x-width y-height t)))
(defun open-scratch-frame (&optional fn)
  "Opens the org-capture window in a floating frame that cleans itself up once
you're done. This can be called from an external shell script."
  (let* ((frame-title-format "")
         (preframe (cl-loop for frame in (frame-list)
                            if (equal (frame-parameter frame 'name) "scratch")
                            return frame))
         (frame (unless preframe
                  (make-frame `((name . "scratch")
                                (width . 90)
                                (height . 26)
                                (transient . t)
                                (internal-border-width . 10)
                                (left-fringe . 0)
                                (right-fringe . 0)
                                (undecorated . t)
                                ,(if IS-LINUX '(display . ":0")))))))
    (select-frame-set-input-focus (or frame preframe))
    (when frame
      (with-selected-frame frame
        (if fn
            (call-interactively fn)
          (with-current-buffer (switch-to-buffer "*scratch*")
            (text-scale-set 0.2)
            (when (eq major-mode 'fundamental-mode)
              (emacs-lisp-mode))
            (redisplay)))))))

;; (add-hook 'after-make-frame-functions
;;           (lambda (frame)
;;             (pcase (frame-parameter nil 'name)
;;               ("scratch" (init-default-frame frame))
;;               ("org-protocol" (init-default-frame frame 0.618 0.618)))))
#+end_src
* Set Font
Doom exposes five (optional) variables or controlling fonts in Doom.
+ ~doom-font~ : set default font
+ ~doom-variable-pitch-font~: Set Serif font
+ ~doom-big-font~: used for `doom-big-font-mode`; use this for presentations or streaming
+ ~doom-unicode-font~: Fallback font for Unicode glyphs
#+begin_src emacs-lisp :tangle "config.el"
(unless (bound-and-true-p using-nix-config-p)
  (defun my/font-installed-p (font-name)
    "Check if font with FONT-NAME is available."
    (if (find-font (font-spec :name font-name)) t nil))
  (defadvice! my/use-default-font-a (&rest _)
    "Set `doom-font'!"
    :before #'doom-init-fonts-h
    (cl-loop for font in '( "JetBrains Mono" "Fira Code" "Cascadia Code"
                            "Source Code Pro" "Menlo" "Fantasque Sans Mono"
                            "monospace")
             when (my/font-installed-p font)
             return (setq doom-font (font-spec :family font :size 12)))
    (advice-remove #'doom-init-fonts-h #'my/use-default-font-a))

  ;; use emoji color font
  ;; see @https://emacs-china.org/t/emacs-cairo/9437/13
  (defun my/walle-ui-display-color-emoji? ()
    "Return non-nil if emacs can display color emoji.

Notice that this function assume you have graphics display"
    (or (and EMACS27+
             (featurep 'cairo)
             (let ((frame-font-backend
                    (frame-parameter (selected-frame) 'font-backend)))
               (when (or (memq 'ftcr frame-font-backend)
                         (memq 'ftcrhb frame-font-backend))
                 t)))
        (featurep 'cocoa)))

  (defadvice! my/use-chinese-font-a (&rest _)
    "Set Chinese fonts and Color Emoji fonts."
    :after-until #'doom-init-extra-fonts-h
    (when (and (my/walle-ui-display-color-emoji?) IS-LINUX)
      (cl-loop for font in '("Apple Color Emoji" "Segoe UI Emoji"
                             "Noto Color Emoji" "JoyPixels")
               when (my/font-installed-p font)
               return (set-fontset-font t 'symbol font)))
    (cl-loop for font in '("Adobe Heiti Std" "STXihei" "Microsoft Yahei"
                           "Hiragino Sans GB W6" "WenQuanYi Micro Hei Mono")
             when (my/font-installed-p font)
             ;; return (set-fontset-font t '(#x4e00 . #x9fff) font)
             return (dolist (charset '(kana han cjk-misc bopomofo))
                      (set-fontset-font t charset font)))))

;; 等高配置
(setq face-font-rescale-alist
      '(("Sarasa Mono SC" . 0.99)
        ("Microsoft Yahei" . 1.2)
        ("WenQuanYi Micro Hei Mono" . 0.99)))
#+end_src

* Theme
There are two ways to load a theme. Both assume the theme is installed and
available. You can either set `doom-theme' or manually load a theme with the
`load-theme' function. This is the default:
#+begin_src emacs-lisp :tangle "config.el"
;; (setq doom-theme 'doom-moonlight)
#+end_src

* company
** company-tabnine
#+begin_src emacs-lisp :tangle "packages.el"
(package! company-tabnine)
#+end_src
#+begin_src emacs-lisp :tangle "config.el"
(use-package! company-tabnine
  :when (featurep! :completion company)
  :config
  (setq company-tabnine--disable-next-transform nil)
  (defun my-company--transform-candidates-a (func &rest args)
    (if (not company-tabnine--disable-next-transform)
        (apply func args)
      (setq company-tabnine--disable-next-transform nil)
      (car args)))

  (defun my-company-tabnine-a (func &rest args)
    (when (eq (car args) 'candidates)
      (setq company-tabnine--disable-next-transform t))
    (apply func args))

  (advice-add #'company--transform-candidates :around #'my-company--transform-candidates-a)
  (advice-add #'company-tabnine :around #'my-company-tabnine-a)

  (setq company-idle-delay 0.2) ; Trigger completion immediately.
  (setq company-show-numbers t) ; Number the candidates (use M-1, M-2 etc to select completions).

  ;; Use the tab-and-go frontend.
  (company-tng-configure-default)
  (setq company-frontends
        '(company-tng-frontend
          company-pseudo-tooltip-frontend
          company-echo-metadata-frontend)))
#+end_src
* General configuration
#+begin_src emacs-lisp :tangle "config.el"
;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

(setq word-wrap-by-category t)  ; Chinese Wrap
(setq confirm-kill-emacs nil)   ; kill emacs, not prompt
(setq confirm-kill-processes nil) ; 退出后自动杀掉进程

;; open file in Dotfiles
(defun my/find-file-in-dotfiles ()
  "Browse your `DOTFILES'."
  (interactive)
  (let* ((dotfiles (or (bound-and-true-p mydotfile)
                    (getenv "DOTFILES")
                    (expand-file-name "~/.dotfiles"))))
    (unless (file-directory-p dotfiles)
      (make-directory dotfiles t))
    (doom-project-find-file dotfiles)))
;; open file in org-directory
(defun my/find-file-in-org-directory ()
  "Browse Your `ORG-DIRECTORY'."
  (interactive)
  (let ((org-dir (or (bound-and-true-p org-directory)
                     (expand-file-name "~/org"))))
    (unless (file-directory-p org-dir)
      (make-directory org-dir t))
    (doom-project-find-file org-dir)))
(map! :leader
      (:prefix-map ("d" . "open prjs")
       :desc "dotfile" "d" #'my/find-file-in-dotfiles
       :desc "org-dird" "o" #'my/find-file-in-org-directory))


;; When using the parameter -t emacs start, remove the default title
(add-hook! doom-after-init-modules
  (let ((title (assoc 'title default-frame-alist)))
    (delq! title default-frame-alist)))
#+end_src

* org
Org-mode is one of my favorite features in Emacs.
#+begin_src emacs-lisp :tangle "packages.el"
(package! org-cliplink)
(package! valign)
(package! org-roam-server)
#+end_src
#+begin_src emacs-lisp :tangle "config.el"
;; Set `org-directory'
(setq org-directory (if (not IS-WINDOWS) "~/projects/org")
      org-archive-location (concat org-directory "/.archive/%s::"))

(setq org-ellipsis " [...]" ; Hide the logo
      org-priority-highest ?A
      org-priority-lowest ?E)
#+end_src
**  org-superstar
The header of the beautification org-mdoe
#+begin_src emacs-lisp :tangle "config.el"
(setq org-superstar-headline-bullets-list '("◉" "○" "✸" "✿" "✤" "✜" "◆" "▶")
      ;; org-superstar-headline-bullets-list '("Ⅰ" "Ⅱ" "Ⅲ" "Ⅳ" "Ⅴ" "Ⅵ" "Ⅶ" "Ⅷ" "Ⅸ" "Ⅹ")
      org-superstar-prettify-item-bullets t)
#+end_src
** org-roam
Org-Roam is one of the best notes notes.It is one of the ways to implement
Roam Research. Look at [[https://www.youtube.com/playlist?list=PLwXSqDdn_CpE934BjXMgmzHnlwXMy41TC][this video series]] for more info.
#+begin_src emacs-lisp :tangle "config.el"
;; not open the org-roam buffer when opening an org roam file
(setq +org-roam-open-buffer-on-find-file nil)
;; see @https://www.zmonster.me/2020/06/27/org-roam-introduction.html
(setq org-roam-capture-templates
      '(( "d" "default" plain (function org-roam-capture--get-point)
          "%?"
          :file-name "%<%Y%m%d%H%M%S>-${slug}"
          :head "#+title: ${title}\n#+roam_tags:\n\n")
        ( "a" "alias" plain (function org-roam-capture--get-point)
          "%?"
          :file-name "%<%Y%m%d%H%M%S>-${slug}"
          :head "#+title: ${title}\n#+roam_alias:\n\n")))

(after! org-roam
  (add-to-list 'org-roam-capture-ref-templates
               '("a" "Annotation" plain (function org-roam-capture--get-point)
                 "%U ${body}\n"
                 :file-name "${slug}"
                 :head "#+title: ${title}\n#+roam_key: ${ref}\n#+roam_alias:\n"
                 :immediate-finish t
                 :unnarrowed t)))

;; org-roam-server
(use-package! org-roam-server
  :after org-roam
  :commands (my/org-roam-server-open)
  :init
  (map! :leader
        :desc "org-roam-server" "on" #'my/org-roam-server-open)
  :config
  (setq org-roam-server-host "127.0.0.1"
        org-roam-server-port 9091
        org-roam-server-export-inline-images t
        org-roam-server-authenticate nil
        org-roam-server-network-label-truncate t
        org-roam-server-network-label-truncate-length 60
        org-roam-server-network-label-wrap-length 20)
  (defun my/org-roam-server-open ()
    "Ensure the server is active, then open the roam graph."
    (interactive)
    (let ((sp-mode-p (bound-and-true-p smartparens-global-mode)))
      (and sp-mode-p (smartparens-global-mode -1))
      (org-roam-server-mode 1)
      (and sp-mode-p (smartparens-global-mode 1))
      (browse-url-xdg-open (format "http://localhost:%d" org-roam-server-port)))))
#+end_src
** org-cliplink
#+begin_src emacs-lisp :tangle "config.el"
(use-package! org-cliplink
  :after org
  :config
  (map! :map org-mode-map :localleader "l h" #'org-cliplink))
#+end_src
** valign
Form pixel alignment,Valign-Mode and Display-Line-Mode conflicts,
so turn off ~display-line-numbers-mode~,
How to turn off display-line-numbers-mode see
@https://github.com/hlissner/doom-emacs/issues/870
@https://github.com/hlissner/doom-emacs/issues/827
@https://emacs-china.org/t/major-mode-smart-parens/15360
#+begin_src emacs-lisp :tangle "config.el"
(use-package! valign
  ;; :commands valign-mode
  :init (setq valign-fancy-bar t)
  :hook ((org-mode markdown-mode)  . my/using-valign-mode)
  :config
  (defun my/using-valign-mode ()
    (display-line-numbers-mode -1)
    (valign-mode +1)))
#+end_src

* Chinese input method
In order to obtain better Chinese input environment.
#+begin_src emacs-lisp :tangle "packages.el"
(package! sis :recipe (:type git
                        :host github
                        :repo "laishulu/emacs-smart-input-source"))
(package! pangu-spacing)
(package! ace-pinyin)
(package! rime :ignore IS-WINDOWS)
#+end_src
** sis
smart change input method(EN or ZH)
see@ https://emacs-china.org/t/os-smart-input-source/13436/726
see@ https://emacs-china.org/t/native-os-sis/14089
#+begin_src emacs-lisp :tangle "config.el"
(use-package! sis
  :after evil
  :config
  (push "M-<SPC>" sis-prefix-override-keys)
  (if (bound-and-true-p using-emacs-rime-p)
      (sis-ism-lazyman-config nil "rime" 'native)
    (sis-ism-lazyman-config "1" "2"
                            (cond ((executable-find"fcitx5") 'fcitx5)
                                  ((executable-find "fcitx") 'fcitx))))

  ;; enable the /cursor color/ mode
  (sis-global-cursor-color-mode t)
  ;; enable the /respect/ mode
  (sis-global-respect-mode t)
  ;; enable the /follow context/ mode for all buffers
  (sis-global-context-mode t)
  ;; enable the /inline english/ mode for all buffers
  (sis-global-inline-mode t))
#+end_src
** emacs-rime
RIME Emacs Front End Framework
#+begin_src emacs-lisp :tangle "config.el"
(use-package! rime
  :defer 0.5
  :init
  (if (bound-and-true-p using-nix-config-p)
      (setq rime-disable-predicates '(rime-predicate-evil-mode-p
                                      rime-predicate-after-alphabet-char-p
                                      rime-predicate-prog-in-code-p))
    (setq rime-user-data-dir (expand-file-name "~/.config/fcitx/rime")))
  (setq default-input-method "rime")
  (setq rime-show-candidate 'posframe)

  ;; (setq rime-inline-ascii-trigger 'shift-l)
  :bind
  ;; C-\ to toggle-input-method
  ;; C-` to toggle
  ;; , and . to page up and down
  (:map rime-mode-map
   ;; open rime menu
   ("C-`" . 'rime-send-keybinding))
  (:map rime-active-mode-map
   ("C-j" . 'rime-inline-ascii))

config/doom/config.el
;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!

(setq confirm-kill-processes nil) ; 退出后自动杀掉进程
;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Shanyou Li"
      user-mail-address "shanyouli6@gmail.com")

;;; font
;; Doom exposes five (optional) variables or controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
(defun my/font-installed-p (font-name)
  "Check if font with FONT-NAME is available."
  (if (find-font (font-spec :name font-name)) t nil))
;; (setq doom-unicode-font nil)
(defadvice! my/use-default-font-a (&rest _)
  "Set `doom-font'!"
  :before #'doom-init-fonts-h
  (progn
    (cl-loop for font in '("JetBrains Mono" "Fantasque Sans Mono"
                           "Cascadia Code" "Source Code Pro" "Menlo"
                           "DejaVu Sans Mono" "monospace")
             when (my/font-installed-p font)
             return (setq doom-font (font-spec :family font :size 12)))
    (advice-remove #'doom-init-fonts-h #'my/use-default-font-a)))

;; use emoji color font
;; see @https://emacs-china.org/t/emacs-cairo/9437/13
(defun my/walle-ui-display-color-emoji? ()
  "Return non-nil if emacs can display color emoji.

Notice that this function assume you have graphics display"
  (or (and EMACS27+
           (featurep 'cairo)
           (let ((frame-font-backend
                  (frame-parameter (selected-frame) 'font-backend)))
             (when (or (memq 'ftcr frame-font-backend)
                       (memq 'ftcrhb frame-font-backend))
               t)))
      (featurep 'cocoa)))

(defadvice! my/use-color-emoji-a (&rest _)
  "If you can use a color Emoji, it will be used."
  :after-until #'doom-init-extra-fonts-h
  (when (and (my/walle-ui-display-color-emoji?) IS-LINUX)
    (set-fontset-font t 'symbol "Noto Color Emoji")))

(defadvice! my/use-chinese-font-a (&rest _)
  "Set Chinese fonts."
  :after #'doom-init-extra-fonts-h
  (cl-loop for font in '("zhHei" "Adobe Heiti Std" "STXihei"
                         "WenQuanYi Micro Hei Mono")
           when (my/font-installed-p font)
           return (set-fontset-font t '(#x4e00 . #x9fff)
                                    (font-spec :family font))))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-moonlight)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory
      (cond (IS-LINUX
             (cl-some (lambda (dir)
                        (if (file-directory-p dir)
                            dir))
                      (list
                       "/data/Documents/Org"
                       "~/Org"
                       "~/org"
                       "~/Documents/Org"
                       )))))


(defun my/find-file-in-dotfiles ()
  "Browse your `DOTFILES'."
  (interactive)
  (let* ((dotfiles (or (getenv "DOTFILES")
                       (expand-file-name "~/.dotfiles"))))
    (unless (file-directory-p dotfiles)
      (make-directory dotfiles t))
    (doom-project-find-file dotfiles)))
;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

(setq word-wrap-by-category t) ;; Chinese Wrap
;; kill emacs, not prompt
(setq confirm-kill-emacs nil)
;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c g k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c g d') to jump to their definition and see how
;; they are implemented.

;; Additional recentf package configuration

;; When using the parameter -t emacs start, remove the default title
(add-hook! doom-after-init-modules
  (let ((title (assoc 'title default-frame-alist)))
    (delq! title default-frame-alist)))

(setq recentf-max-saved-items 300
      recentf-exclude
      '("\\.?cache" ".cask" "url" "COMMIT_EDITMSG\\'" "bookmars"
        "\\.\\(?.gz\\|gif\\|svg\\|png\\|jpe?g\\|bmp\\|xpm\\)$"
        "\\.?ido\\.last$" "\\.revive$" "/G?TAGS$" "/.elfeed/"
        "^/tmp/" "^/var/folders/.+$"
        (lambda (file)
          (or (file-directory-p file)
              (file-in-directory-p file package-user-dir)
              (and (bound-and-true-p straight-base-dir)
                   (file-in-directory-p
                    file
                    (concat straight-base-dir "straight")))
              (and (bound-and-true-p persp-save-dir)
                   (file-in-directory-p file persp-save-dir))
              (and (bound-and-true-p desktop-dirname)
                   (file-in-directory-p file desktop-dirname))))))
(after! org-roam
  (add-to-list 'org-roam-capture-ref-templates
               '("a" "Annotation" plain (function org-roam-capture--get-point)
                 "%U ${body}\n"
                 :file-name "${slug}"
                 :head "#+title: ${title}\n#+roam_key: ${ref}\n#+roam_alias:\n"
                 :immediate-finish t
                 :unnarrowed t)))

(after! org-superstar
  (setq org-superstar-headline-bullets-list '("◉" "○" "✸" "✿" "✤" "✜" "◆" "▶")
        ;; org-superstar-headline-bullets-list '("Ⅰ" "Ⅱ" "Ⅲ" "Ⅳ" "Ⅴ" "Ⅵ" "Ⅶ" "Ⅷ" "Ⅸ" "Ⅹ")
        org-superstar-prettify-item-bullets t ))
(after! org
  (setq org-ellipsis "▼"
        org-priority-highest ?A
        org-priority-lowest ?E
        org-priority-faces
        '((?A . 'all-the-icons-red)
          (?B . 'all-the-icons-orange)
          (?C . 'all-the-icons-yellow)
          (?D . 'all-the-icons-green)
          (?E . 'all-the-icons-blue))))
(after! recentf
  (push (expand-file-name recentf-save-file) recentf-exclude))

(after! all-the-icons
  (push '("\\.ass\\'" all-the-icons-material "subtitles"
          :face all-the-icons-yellow)
        all-the-icons-icon-alist))

(use-package! all-the-icons-ivy-rich
  :if (and (featurep! :completion ivy) (featurep! :emacs dired +icons))
  :hook (ivy-mode . all-the-icons-ivy-rich-mode))

(use-package! eaf
  :commands eaf-open
  :init
  (setq eaf-config-location (concat doom-local-dir "eaf/"))
  (setq eaf-proxy-type "http"
        eaf-proxy-host "127.0.0.1"
        eaf-proxy-port "7890")
  (setq eaf-find-alternate-file-in-dired t)
  :hook (eaf-mode . evil-emacs-state)
  :config
  ;; (eaf-setq eaf-browse-blank-page-url "https://duckduckgo.com")
  (setq eaf-browser-default-search-engine "duckduckgo"))

(when (featurep! :checkers syntax)
  (remove-hook 'doom-first-buffer-hook 'global-flycheck-mode)
  (add-hook 'sh-mode-hook 'flycheck-mode))

(use-package! rime
  :after-call after-find-file pre-command-hook
  :custom (default-input-method "rime")
  :init
  (setq rime-user-data-dir (concat doom-local-dir "rime/"))
  :config
  (unless rime-emacs-module-header-root
    (setq-default rime-emacs-module-header-root
                  (concat "/usr/include/" "emacs-"
                          (number-to-string emacs-major-version)
                          (if (eq emacs-minor-version 0)
                              "-vcs")
                          "/")))
  (when (and (display-graphic-p) (require 'posframe nil t))
    (setq rime-show-candidate 'posframe)
    ;;rime-posframe-style (list :background-color "#333333"
    ;;                        :foreground-color "#dcdccc"
    ;;                          :font "Sarasa Mono SC"
    ;;                          :internal-border-width 10))
    ;; (setq rime-disable-predicates
    ;;       '(rime-predicate-evil-mode-p
    ;;         rime-predicate-after-alphabet-char-p
    ;;         rime-predicate-prog-in-code-p))
    (setq rime-posfrmae-style 'horizontal)))

(use-package! sis
  :after evil
  :config
  (push "M-<SPC>" sis-prefix-override-keys)
  (sis-ism-lazyman-config "1" "2"
                          (cond ((executable-find"fcitx5") 'fcitx5)
                                ((executable-find "fcitx") 'fcitx)))
  ;; enable the /cursor color/ mode
  (sis-global-cursor-color-mode t)
  ;; enable the /respect/ mode
  (sis-global-respect-mode t)
  ;; enable the /follow context/ mode for all buffers
  (sis-global-follow-context-mode t)
  ;; enable the /inline english/ mode for all buffers
  (sis-global-inline-mode t))

(use-package! pangu-spacing
  :hook (text-mode . pangu-spacing-mode)
  :config
  ;; Always insert `real' space in org-mode.
  (setq-hook! 'org-mode-hook pangu-spacing-real-insert-separtor t))

(use-package! ace-pinyin
  :after avy
  :init (setq ace-pinyin-use-avy t)
  :config (ace-pinyin-global-mode 1))

(defadvice! +chinese--org-html-paragraph-a (args)
  "Join consecutive Chinese lines into a single long line without unwanted space
when exporting org-mode to html."
  :filter-args #'org-html-paragraph
  (cl-destructuring-bind (paragraph contents info) args
    (let* ((fix-regexp "[[:multibyte:]]")
           (fixed-contents
            (replace-regexp-in-string
             (concat "\\(" fix-regexp "\\) *\n *\\(" fix-regexp "\\)")
             "\\1\\2"
             contents)))
      (list paragraph fixed-contents info))))

(use-package! sdcv
  :commands lye-dict-search-at-point
  :init
  (setq sdcv-dictionary-data-dir (expand-file-name "sdcv"
                                                   (or (getenv "XDG_DATA_HOME")
                                                       "~/.local/share")))
  :config
  (setq sdcv-dictionary-simple-list ; setup dictionary list for simple search
        '("KDic11万英汉词典"
          "懒虫简明英汉词典"
          "懒虫简明汉英词典"))
  (setq sdcv-dictionary-complete-list ; setup dictionary list for complete search
        '("KDic11万英汉词典"
          "懒虫简明英汉词典"
          "懒虫简明汉英词典"
          "21世纪英汉汉英双向词典"
          "新世纪汉英科技大词典"
          "牛津现代英汉双解词典"
          "XDICT汉英辞典"
          "XDICT英汉辞典"
          "朗道汉英字典5.0"
          "朗道英汉字典5.0"
          "quick_eng-zh_CN"
          "CDICT5英汉辞典"))
  (defun lye-dict-search-at-point ()
    (interactive)
    (if (display-graphic-p)
        (call-interactively #'sdcv-search-pointer+)
      (call-interactively #'sdcv-search-pointer))))

(use-package! youdao-dictionary
  :commands lye-dict-search-at-point
  :init
  (setq youdao-dictionary-search-history-file ; Set file for saving search history
        (concat doom-local-dir "ydcv/"))
  :config
  (setq url-automatic-caching t
        ;; Enable Chinese word segmentation support (支持中文分词)
        youdao-dictionary-use-chinese-word-segmentaton t)
  (defun lye-dict-search-at-point ()
    (interactive)
    (if (display-graphic-p)
        (call-interactively #'youdao-dictionary-search-at-point-posframe)
      (call-interactively #'youdao-dictionary-search-at-point))))

(use-package! english-teacher
  :hook ((Info-mode
          elfeed-show-mode
          eww-mode
          Man-mode
          Woman-Mode) . english-teacher-follow-mode))

(use-package! super-save
  :hook (doom-first-file . super-save-mode)
  :config
  (push 'split-window-below super-save-triggers)
  (push 'split-window-right super-save-triggers)
  (push 'aw--select-window super-save-triggers)
  (push 'ace-window super-save-triggers))

(use-package! insert-translated-name
  :commands (insert-translated-name-insert-original-translation
             insert-translated-name-insert-with-underline
             insert-translated-name-insert-with-line
             insert-translated-name-insert-with-camel
             insert-translated-name-insert)
  :config
  (defvar int--evil-last-status-is-insert-p nil
      "Last evil-mode status is inserted?")
  (defvar int--sis-default-input-thmod nil
      "Staging the default input method.")
  (defun int/active-a (&rest _)
    (when (and (fboundp 'evil-insert-state-p) (evil-insert-state-p))
      (setq int--evil-last-status-is-insert-p t))
    (unless int--evil-last-status-is-insert-p
      (evil-insert-state))
    (when (boundp 'int--sis-default-input-thmod)
      (setq int--sis-default-input-thmod default-input-method)
      (setq-local default-input-method nil)
      (sis-set-other)))

  (defun int/inactive-a (&rest _)
    (unless int--evil-last-status-is-insert-p
      (evil-normal-state)
      (setq int--evil-last-status-is-insert-p nil))
    (when (bound-and-true-p int--sis-default-input-thmod)
      (sis-set-english)
      (setq-local default-input-method int--sis-default-input-thmod)
      (setq int--sis-default-input-thmod nil)))
  (advice-add #'insert-translated-name-active :before 'int/active-a)
  (advice-add #'insert-translated-name-inactive :after 'int/inactive-a))

(use-package! rainbow-mode
  :hook (prog-mode . rainbow-mode))

(use-package! css-mode
  :mode ("\\.rasi\\'" . css-mode))

(use-package! hungry-delete
  :hook (doom-first-buffer . global-hungry-delete-mode))

(when (executable-find "vterm-ctrl")
  (after! vterm
    (setq vterm-module-cmake-args "-DUSE_SYSTEM_LIBVTERM=yes")))

(defadvice! +no-query-kill-emacs-a (orign &rest args)
  "Prevent annoying \"Active process exits\" query when you quit Emacs."
  :around #'save-buffers-kill-emacs
  (require 'noflet)
  (noflet ((process-list ()))
          (apply orign args)))

(use-package! valign
  ;; :commands valign-mode
  :hook ((org-mode markdown-mode)  . valign-mode))

(use-package! sh-script
  :mode (("/.bashrc\\'" . sh-mode)))

(use-package! gif-screencast
  :commands gif-screencast-start-or-stop
  :config
  (setq gif-screencast-output-directory (expand-file-name "~/Videos/Gif")))

(use-package! org-roam-server
  :after org-roam
  :config
  (setq org-roam-server-host "127.0.0.1"
        org-roam-server-port 9091
        org-roam-server-export-inline-images t
        org-roam-server-authenticate nil
        org-roam-server-network-label-truncate t
        org-roam-server-network-label-truncate-length 60
        org-roam-server-network-label-wrap-length 20)
  (defun org-roam-server-open ()
    "Ensure the server is active, then open the roam graph."
    (interactive)
    (org-roam-server-mode 1)
    (browse-url-xdg-open (format "http://localhost:%d" org-roam-server-port))))

;;; Window-select
(use-package! ace-window
  :bind ([remap other-window] . ace-window)
  :hook (emacs-startup . ace-window-display-mode)
  :config
  ;; Select window via 'C-w 1' ... `C-w 9'
  (defun aw--select-window (number)
    "Slecet the specified window."
    (when (numberp number)
      (let ((found nil))
        (dolist (win (aw-window-list))
          (when (and (window-live-p win)
                     (eq number (string-to-number (window-parameter win 'ace-window-path))))
            (setq found t)
            (aw-switch-to-window win)))
        (unless found
          (message "No specified window: %d" number)))))
  (defun aw--select-window-1 ()
    (interactive )
    (aw--select-window 1))
  (defun aw--select-window-2 ()
    (interactive )
    (aw--select-window 2))
  (defun aw--select-window-3 ()
    (interactive )
    (aw--select-window 3))
  (defun aw--select-window-4 ()
    (interactive )
    (aw--select-window 4))
  (defun aw--select-window-5 ()
    (interactive )
    (aw--select-window 5))
  (defun aw--select-window-6 ()
    (interactive )
    (aw--select-window 6))
  (defun aw--select-window-7 ()
    (interactive )
    (aw--select-window 7))
  (defun aw--select-window-8 ()
    (interactive )
    (aw--select-window 8))
  (defun aw--select-window-9 ()
    (interactive )
    (aw--select-window 9))
  ;; (dotimes (n 9)
  ;;   (bind-key (format "M-%d" (1+ n))
  ;;             (lambda ()
  ;;               (interactive)
  ;;               (aw--select-window (1+ n))))
  (map! :map evil-window-map
        "1" #'aw--select-window-1
        "2" #'aw--select-window-2
        "3" #'aw--select-window-3
        "4" #'aw--select-window-4
        "5" #'aw--select-window-5
        "6" #'aw--select-window-6
        "7" #'aw--select-window-7
        "8" #'aw--select-window-8
        "9" #'aw--select-window-9)
  (map! "M-1" #'aw--select-window-1
        "M-2" #'aw--select-window-2
        "M-3" #'aw--select-window-3
        "M-4" #'aw--select-window-4
        "M-5" #'aw--select-window-5
        "M-6" #'aw--select-window-6
        "M-7" #'aw--select-window-7
        "M-8" #'aw--select-window-8
        "M-9" #'aw--select-window-9))

(use-package! org-cliplink
  :after org
  :config
  (map! :map org-mode-map
        :localleader
        "l h" #'org-cliplink))
;;; Key
(map! :leader
      (:prefix-map ("e" . "eaf")
       (:when (fboundp 'eaf-open)
        :desc "Open a terimal" "t" #'eaf-open-terminal
        :desc "Open a file" "o" #'eaf-open
        :desc "Open a url" "l" #'eaf-open-url
        :desc "Search it" "s" #'eaf-search-it))
      (:when (fboundp 'lye-dict-search-at-point)
       :desc "Dictionary" "y" #'lye-dict-search-at-point)
      (:prefix-map ("i" . "insert")
       (:when (fboundp 'insert-translated-name-insert)
        :desc "Chinese to English" "t" #'insert-translated-name-insert-original-translation
        (:prefix-map ("T" . "ZH-EN format")
         :desc "underline" "u" #'insert-translated-name-insert-with-underline
         :desc "Original" "o" #'insert-translated-name-insert-original-translation
         :desc "line" "l" #'insert-translated-name-insert-with-line
         :desc "Camel" "c" #'insert-translated-name-insert-with-camel
         :desc "intelligent" "i" #'insert-translated-name-insert)))
      (:prefix-map ("u" . "user keys")
       (:when (fboundp 'gif-screencast-start-or-stop)
        :desc "gif screencast" "g" #'gif-screencast-start-or-stop)))
;; (map! "<XF86Tools>" 'toggle-input-method
;;       (:after rime
;;         :map rime-mode-map
;;         "C-`" #'rime-send-keybinding))

